python-gitlab Issue #63 - implement pagination for list()

jantman opened the pull request:
This implements iteration/pagination of list() results for #63

gpocentek commented on gitlab/__init__.py:
```diff
@@ -26,6 +26,9 @@
 import six
 
+import logging
+logger = logging.getLogger(__name__)
```
Could you remove the logging code for this change? I'm not against adding some logging but I'd rather do it in another patch.

jantman replied:
Done.

gpocentek commented on gitlab/__init__.py:
```diff
@@ -189,6 +192,8 @@ def set_url(self, url):
         self._url = '%s/api/v3' % url
 
     def _construct_url(self, id_, obj, parameters):
+        if 'next_url' in parameters:
```
Could you handle prev, first and last as well?

jantman replied:
@gpocentek I'm not sure I understand the purpose of that, given how this function is called. Even if I don't use the generator idea, from @dserodio, it's your project but I'd rather not add in additional unused code paths...

gpocentek replied:
My comment was just plain stupid. Sorry.

gpocentek commented on gitlab/__init__.py:
```diff
@@ -342,8 +347,14 @@ def list(self, obj_class, **kwargs):
+                args = kwargs.copy()
+                args['next_url'] = r.links['next']['url']
+                logger.debug("Iterating results 'next' link: %s", args['next_url'])
+                results.extend(self.list(obj_class, **args))
```
This doesn't seem to give a choice to the user, this will always return the full list and do multiple API calls. If a user has hundreds of items this is a problem. It should be optional.

dserodio replied:
IMHO, the perfect implementation would use Python generators instead of a list, and only fetch the next page when needed.

gpocentek replied:
Sounds like a good plan. I'm a bit short on time to work on this these days but I'd be happy to merge this solution. Thanks!

jantman replied:
I'm not sure I see this as "a problem" - the whole reason why I opened this PR is because I have about 200 users, and will soon have about 300 repositories. I've seen some latency when fetching all that, but not much.

I like @dserodio's idea of using a generator for this. I'm going to have a try at implementing it, though I haven't done much with generators before, and perhaps worse, I've never really worked with ``unittest`` or ``httmock``.

jantman commented:
I've removed the logging code as requested, and added unit tests for the _current_ implementation of just the ``next`` link. However, I think I've hit a bit of a roadblock...

I have nothing against giving a choice to the user, but I'd much rather prefer the iteration to be the default, with disabling it an option. I guess this is personal preference, but I find it very annoying when an API client presents a method like ``Users()`` that only retrieves partial results without making that extremely clear.

I looked into turning this into a generator, but as far as I can tell, I'd need to drastically change how arguments and return values are passed around to get away from the current recursive pattern.

gpocentek commented:
Thanks Jason.

I've merged your patch as is but I'll probably restore the default behavior to avoid surprises on updates (people will not expect to suddenly retrieve all the items).

jantman commented:
Ok, thanks! I'll keep an eye out for it, and update my code to use the pagination once it's released.
